src/ionosphere.c

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2015 Swift Navigation Inc.
+ * Contact: Leith Bade <leith@swift-nav.com>
+ *
+ * This source is subject to the license found in the file 'LICENSE' which must
+ * be be distributed together with this source. All other rights reserved.
+ *
+ * THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+ * EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+ */
+
+#include <float.h>
+#include <math.h>
+#include <stdio.h>
+
+#include "constants.h"
+#include "ionosphere.h"
+
+/** \defgroup ionosphere Ionospheric models
+ * Implemenations of ionoshperic delay correction models.
+ * \{ */
+
+/** Calculate ionospheric delay using Klobuchar model.
+ *
+ * References:
+ *   -# IS-GPS-200H, Section 20.3.3.5.2.5 and Figure 20-4
+ *
+ * \param t_gps GPS time at which to calculate the ionospheric delay
+ * \param lat_u Latitude of the receiver [rad]
+ * \param lon_u Longitude of the receiver [rad]
+ * \param a Azimuth of the satellite, clockwise positive from North [rad]
+ * \param e Elevation of the satellite [rad]
+ * \param i Ionosphere parameters struct from GPS NAV data
+ *
+ * \return  Ionospheric delay distance for GPS L1 frequency [m]
+ */
+double calc_ionosphere(gps_time_t t_gps,
+                       double lat_u, double lon_u,
+                       double a, double e,
+                       const ionosphere_t *i)
+{
+  /* Convert inputs from radians to semicircles */
+  /* All calculations are in semicircles */
+  lat_u = lat_u / M_PI;
+  lon_u = lon_u / M_PI;
+  /* a can remain in radians */
+  e = e / M_PI;
+
+  /* Calculate the earth-centered angle */
+  double psi = 0.0137 / (e + 0.11) - 0.022;
+
+  /* Compute the latitude of the Ionospheric Pierce Point */
+  double lat_i = lat_u + psi * cos(a);
+  if (lat_i > 0.416) {
+    lat_i = 0.416;
+  }
+  if (lat_i < -0.416) {
+    lat_i = -0.416;
+  }
+
+  /* Compute the longitude of the IPP */
+  double lon_i = lon_u + (psi * sin(a)) / cos(lat_i * M_PI);
+
+  /* Find the geomagnetic latitude of the IPP */
+  double lat_m = lat_i + 0.064 * cos((lon_i - 1.617) * M_PI);
+
+  /* Find the local time at the IPP */
+  double t = 43200.0 * lon_i + t_gps.tow;
```

Comment:
As per chat, `normalize_gps_time` normalises to a week. This model needs to be normalised to a day.